### ISSUE TITLE
ComputeError when using lazy pl.int_range() in combination with pl.len() in with_columns(), but not in select(). 

### ISSUE THREAD

**Issue by Tachy0o-n** [pola-rs/polars#17424]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
lf = pl.LazyFrame({'a' : [0,0,0]}) 

(lf.with_columns(pl.int_range(start = i,
                              end = i+pl.col('a').len())
                   .alias(str(i)) for i in range(2))
   .collect())

# ComputeError: `end` must contain exactly one value, got 3 values
```

The workaround is using select().

```python
(lf.select(pl.all(),
           *(pl.int_range(start = i,
                          end = i+pl.col('a').len())
                   .alias(str(i)) for i in range(2)))
   .collect())

# shape: (3, 3)
# ┌─────┬─────┬─────┐
# │ a   ┆ 0   ┆ 1   │
# │ --- ┆ --- ┆ --- │
# │ i64 ┆ i64 ┆ i64 │
# ╞═════╪═════╪═════╡
# │ 0   ┆ 0   ┆ 1   │
# │ 0   ┆ 1   ┆ 2   │
# │ 0   ┆ 2   ┆ 3   │
# └─────┴─────┴─────┘
```

### Issue description

The above MWE works when using...
- a DataFrame instead of a LazyFrame.
- the value 3 instead of pl.col('a').len().
- a fixed value instead of i in start/end.

### Expected behavior

The lazy with_columns() version should work.

### Installed versions

<details>

```
--------Version info---------
Polars:               1.0.0
Index type:           UInt32
Platform:             Linux-5.15.0-113-generic-x86_64-with-glibc2.35
Python:               3.11.7 | packaged by conda-forge | (main, Dec 23 2023, 14:43:09) [GCC 12.3.0]

----Optional dependencies----
adbc_driver_manager:  0.10.0
cloudpickle:          <not installed>
connectorx:           0.3.2
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
great_tables:         <not installed>
hvplot:               <not installed>
matplotlib:           3.8.2
nest_asyncio:         1.6.0
numpy:                1.26.4
openpyxl:             <not installed>
pandas:               2.2.1
pyarrow:              15.0.1
pydantic:             <not installed>
pyiceberg:            <not installed>
sqlalchemy:           2.0.25
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

**Comment by cmdlineluser** [comment]
Looks like it may be a CSE optimization issue:

```python
(lf.with_columns(pl.int_range(start = i,
                              end = i+pl.col('a').len())
                   .alias(str(i)) for i in range(2))
   .collect(comm_subexpr_elim=False))

# shape: (3, 3)
# ┌─────┬─────┬─────┐
# │ a   ┆ 0   ┆ 1   │
# │ --- ┆ --- ┆ --- │
# │ i64 ┆ i64 ┆ i64 │
# ╞═════╪═════╪═════╡
# │ 0   ┆ 0   ┆ 1   │
# │ 0   ┆ 1   ┆ 2   │
# │ 0   ┆ 2   ┆ 3   │
# └─────┴─────┴─────┘
```

**Comment by Tachy0o-n** [comment]
The with_columns() MWE works when using pl.len() instead of pl.col('a').len().

```python
lf = pl.LazyFrame({'a' : [0,0,0]}) 

(lf.with_columns(pl.int_range(start = i,
                              end = i+pl.len())
                   .alias(str(i)) for i in range(2))
   .collect())

# shape: (3, 3)
# ┌─────┬─────┬─────┐
# │ a   ┆ 0   ┆ 1   │
# │ --- ┆ --- ┆ --- │
# │ i64 ┆ i64 ┆ i64 │
# ╞═════╪═════╪═════╡
# │ 0   ┆ 0   ┆ 1   │
# │ 0   ┆ 1   ┆ 2   │
# │ 0   ┆ 2   ┆ 3   │
# └─────┴─────┴─────┘
``` 

But it still failed in my more complex use case, where I need the length of a second LazyFrame. I tried to rewrite my code using reshape, but the same error occurs:
```python
lf = pl.LazyFrame({'a' : [0,0,0]}) 

lf_len = pl.LazyFrame({'x' : [1,1]})

lf = pl.concat([lf, lf_len.select(pl.len().alias('len'))], how='horizontal')

x=2

(lf.with_columns(pl.int_range(start = pl.col('len').first(),
                              end = pl.col('len').first()+x*pl.len())
                   .reshape((-1,x))
                   .arr.to_struct()
                   .alias('struct'))
   .unnest('struct')
   .collect())

# polars.exceptions.ComputeError: `start` must contain exactly one value, got 3 values
```

Without CSE, it works.
```python
...
.collect(comm_subexpr_elim=False))

# shape: (3, 4)
# ┌─────┬──────┬─────────┬─────────┐
# │ a   ┆ len  ┆ field_0 ┆ field_1 │
# │ --- ┆ ---  ┆ ---     ┆ ---     │
# │ i64 ┆ u32  ┆ i64     ┆ i64     │
# ╞═════╪══════╪═════════╪═════════╡
# │ 0   ┆ 2    ┆ 2       ┆ 3       │
# │ 0   ┆ null ┆ 4       ┆ 5       │
# │ 0   ┆ null ┆ 6       ┆ 7       │
# └─────┴──────┴─────────┴─────────┘
```

**Comment by cmdlineluser** [comment]
It does seem to be a regression as it works in `0.20.31`

It will likely need @ritchie46 to take a look.

**Comment by Tachy0o-n** [comment]
This is fixed in version 1.17.1 (#20222).
